usl_pipeline/cloud_functions/main.py

```diff
@@ -903,3 +900,98 @@ def _write_as_npy(blob: storage.Blob, array: NDArray) -> None:
+    """Creates scaled version for every chunk file in the storage bucket parent folder.
+
+    Output files with scaled feature matrices will be placed to the same folder where
+    unscaled files (following "chunk_*" name pattern) are stored but will get "scaled_"
+    prefix. Rescaling only starts when metadata of study area (corresponding to the
+    parent folder name) contains chunks counters giving exactly the same number of
+    chunks as are currently stored in the bucket folder. This should happen when the
+    last feature chunk is stored.
+
+    Args:
+        feature_bucket: Storage bucket with feature matrices.
+        blob_path: Path pointing to one of unscaled feature matrix files with the name
+            following the "chunk_*" pattern. Other files are ignored.
+    """
+    study_area_name, chunk_name = _parse_chunk_path(blob_path)
+    # Let's look up chunk metadata and check if scaling is needed
+    db = firestore.Client()
+    chunk_metadata = metastore.StudyAreaSpatialChunk.get_if_exists(
+        db, study_area_name, chunk_name
+    )
+    if chunk_metadata is None:
+        logging.info(f"Chunk metadata was not found for {study_area_name}/{chunk_name}")
+        return
+
+    if not chunk_metadata.needs_scaling:
+        logging.info(f"Chunk {study_area_name}/{chunk_name} doesn't need scaling")
+        return
+
+    study_area = metastore.StudyArea.get(db, study_area_name)
+    if study_area.chunk_x_count is None or study_area.chunk_y_count is None:
+        logging.info(
+            f"[Feature Rescaler] Study area {study_area_name} has no chunk counts in "
+            + "metadata"
+        )
+        return
+
+    feature_blobs = list(feature_bucket.list_blobs(prefix=f"{study_area_name}/chunk_"))
+
+    chunk_count = study_area.chunk_x_count * study_area.chunk_y_count
+    if chunk_count != len(feature_blobs):
+        logging.info(
+            f"[Feature Rescaler] Study area {study_area_name} has chunk counts"
+            + f" in metadata ({chunk_count}) different from number of stored feature"
+            + f" chunks ({len(feature_blobs)})"
+        )
+        return
+
+    # We're at the final step when all unscaled feature matrices are ready
+    for feature_blob in feature_blobs:
+        logging.info(f"[Feature Rescaler] Rescaling feature matrix {feature_blob.name}")
+        _, blob_chunk_name = _parse_chunk_path(feature_blob.name)
+        output_blob = feature_bucket.blob(
+            f"{study_area_name}/scaled_{blob_chunk_name}.npy"
+        )
+        with feature_blob.open("rb") as in_fd, output_blob.open("wb") as out_fd:
+            feature_raster_transformers.rescale_feature_blob(in_fd, study_area, out_fd)
+
+    logging.info("[Feature Rescaler] Updating feature matrix path in chunk metadata...")
+    for feature_blob in feature_blobs:
+        _, blob_chunk_name = _parse_chunk_path(feature_blob.name)
+        scaled_feature_matrix_path = (
```

Comment:
Should we stick this logic in the loop above? Or did you want to explicitly avoid writing any metadata until all chunks are scaled?

Reply:
Yeah, I'm thinking about retry logic that will try to fix some error if it's related to scaling. If we delete some part of original unscaled files retry logic wouldn't be possible to proceed.

Reply:
I ended up with quite a refactoring. Now the idea is we generate `trigger` files for each feature matrix file. Each trigger-file triggers separate CF processing for corresponding feature matrix.